### PR TITLE
Fix race in Distributed table startup

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix race in Distributed table startup (that could lead to processing file of async INSERT multiple times)

Before this patch it was possible to have multiple directory monitors for the same directory, one from the INSERT context, another one on storage startup().

Here are an example of logs for this scenario:

    2022.12.07 12:12:27.552485 [ 39925 ] {a47fcb32-4f44-4dbd-94fe-0070d4ea0f6b} <Debug> DDLWorker: Executed query: DETACH TABLE inc.dist_urls_in
    ...
    2022.12.07 12:12:33.228449 [ 4408 ] {20c761d3-a46d-417b-9fcd-89a8919dd1fe} <Debug> executeQuery: (from 0.0.0.0:0, user: ) /* ddl_entry=query-0000089229 */ ATTACH TABLE inc.dist_urls_in (stage: Complete)
    ... this is the DirectoryMonitor created from the context of INSERT for the old StoragePtr that had not been destroyed yet (becase of "was 1" this can be done only from the context of INSERT) ...
    2022.12.07 12:12:35.556048 [ 39536 ] {} <Trace> inc.dist_urls_in.DirectoryMonitor: Files set to 173 (was 1)
    2022.12.07 12:12:35.556078 [ 39536 ] {} <Trace> inc.dist_urls_in.DirectoryMonitor: Bytes set to 29750181 (was 71004)
    2022.12.07 12:12:35.562716 [ 39536 ] {} <Trace> Connection (i13.ch:9000): Connected to ClickHouse server version 22.10.1.
    2022.12.07 12:12:35.562750 [ 39536 ] {} <Debug> inc.dist_urls_in.DirectoryMonitor: Sending a batch of 10 files to i13.ch:9000 (0.00 rows, 0.00 B bytes).
    ... this is the DirectoryMonitor that created during ATTACH ...
    2022.12.07 12:12:35.802080 [ 39265 ] {} <Trace> inc.dist_urls_in.DirectoryMonitor: Files set to 173 (was 0)
    2022.12.07 12:12:35.802107 [ 39265 ] {} <Trace> inc.dist_urls_in.DirectoryMonitor: Bytes set to 29750181 (was 0)
    2022.12.07 12:12:35.834216 [ 39265 ] {} <Debug> inc.dist_urls_in.DirectoryMonitor: Sending a batch of 10 files to i13.ch:9000 (0.00 rows, 0.00 B bytes).
    ...
    2022.12.07 12:12:38.532627 [ 39536 ] {} <Trace> inc.dist_urls_in.DirectoryMonitor: Sent a batch of 10 files (took 2976 ms).
    ...
    2022.12.07 12:12:38.601051 [ 39265 ] {} <Error> inc.dist_urls_in.DirectoryMonitor: std::exception. Code: 1001, type: std::__1::__fs::filesystem::filesystem_error, e.what() = filesystem error: in file_size: No such file or directory ["/data6/clickhouse/data/inc/dist_urls_in/shard13_replica1/66827403.bin"], Stack trace (when copying this message, always include the lines below):
    ...
    2022.12.07 12:12:54.132837 [ 4408 ] {20c761d3-a46d-417b-9fcd-89a8919dd1fe} <Debug> DDLWorker: Executed query: ATTACH TABLE inc.dist_urls_in

And eventually both monitors (for a short period of time, one replaces another) are trying to process the same batch (current_batch.txt), and one of them fails because such file had been already removed.

### CODE REVIEW

File: src/Storages/StorageDistributed.cpp [hunk 1]
```diff
@@ -1234,39 +1234,19 @@ StorageDistributedDirectoryMonitor& StorageDistributed::requireDirectoryMonitor(
...
             monitors_blocker,
             getContext()->getDistributedSchedulePool(),
             /* initialize_from_disk= */ startup);
-        return data;
-    };
-
-    /// In case of startup the lock can be acquired later.
-    if (startup)
-    {
-        auto tmp_node_data = create_node_data();
```
Comment (antonio2368) [inline src/Storages/StorageDistributed.cpp:1253]: This was the problematic part as it was done outside of the lock?

**Review by antonio2368 — APPROVED** [review]

File: src/Storages/StorageDistributed.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/StorageDistributed.cpp:1253]: Yes.

**Comment by azat** [comment]
[Stateless tests (release) — Tests are not finished, fail: 0, passed: 0](https://s3.amazonaws.com/clickhouse-test-reports/45360/54fc6859ae7f692308b56a1f94aac6b26841862f/stateless_tests__release_.html)
[Stateless tests (release, wide parts enabled) — Tests are not finished, fail: 0, passed: 0](https://s3.amazonaws.com/clickhouse-test-reports/45360/54fc6859ae7f692308b56a1f94aac6b26841862f/stateless_tests__release__wide_parts_enabled_.html)

>2023-01-18 01:59:14 clickhouse-test: error: unrecognized arguments: --report-logs-stats

Likely the problem is that the image was new, that calls `clickhouse-test` with `--report-logs-stats`, while `clickhouse-test` itself, does not have this option the merge-commit for this PR.

**Comment by antonio2368** [comment]
I'll merge as soon the basic tests pass

**Pull request merged**